File: client/src/lib/backportPack.ts
/** Installing a community "backport pack".
 *
 *  A pack is how backports are actually distributed, and it is a superset of
 *  what the library-set flow does: alongside `fakelib/` it carries an
 *  `eboot.bin` ALREADY patched to the backport SDK pair, plus replacement
 *  `sce_module/` modules. Installing one is therefore a file copy and NOT an
 *  SDK patch — patching a pre-patched eboot would rewrite a field that is
 *  already correct and invalidate the fake-signature the pack shipped with.
 *
 *  This deliberately does not reuse `BackportRecord`/`undoBackport`. That undo
 *  restores `fakelib/` only and then calls `sdk/restore`, which reverts from
 *  the `.bak` files the SDK patch leaves behind — a pack install writes files
 *  outside `fakelib/` and creates no `.bak`, so reusing it would silently fail
 *  to put back the eboot it overwrote.
 */

import type { InstalledTitle } from "../api/ps5";

/** Where per-title backups live. Same root the library flow uses, so a user
 *  looking for "what did this tool leave on my console" finds one place. */
export const PACK_STASH_ROOT = "/data/ps5upload/backport";

export interface PackFile {
  /** Path relative to the pack root, always with `/` separators. */
  relPath: string;
  size: number;
}

export interface BackportPack {
  isPack: boolean;
  /** Title id read out of the folder name. A HINT, never an authority. */
  titleIdHint: string | null;
  libraries: PackFile[];
  eboot: PackFile | null;
  sceModules: PackFile[];
  /** The game's own engine plugins (`prx/`). Title-specific, so never corpus
   *  material — but a pack that ships them expects them installed. */
  gamePrx: PackFile[];
  /** `sce_sys/about/` — present in every pack seen so far. */
  sceSys: PackFile[];
  other: PackFile[];
  totalBytes: number;
}

export interface PackCopy {
  /** Absolute path on the user's computer. */
  from: string;
  /** Absolute path on the console. */
  to: string;
  size: number;
}

export interface PackInstallPlan {
  target: InstalledTitle;
  packRoot: string;
  copies: PackCopy[];
  /** Live console files this install would overwrite, and where each original
   *  gets stashed. Only files that actually exist appear here. */
  stashed: { live: string; stash: string }[];
  stashDir: string;
  /** The pack names a different title than the one selected. Not fatal — the
   *  name is only a folder string — but the user must confirm, because the
   *  alternative is overwriting a 256 MB eboot with another game's. */
  titleMismatch: boolean;
  totalBytes: number;
}

export interface PackInstallRecord {
  titleId: string;
  targetSource: string;
  packRoot: string;
  copiedPaths: string[];
  stashed: { live: string; stash: string }[];
  stashDir: string;
  /** False when apply threw partway. Undo must still be offered: a title with
   *  half a pack installed launches and aborts. */
  complete: boolean;
}

export interface PackTransport {
  mkdirConsole(path: string): Promise<void>;
  copyConsole(from: string, to: string): Promise<void>;
  uploadHost(from: string, to: string): Promise<void>;
  remove(path: string): Promise<void>;
}

export class PackInstallError extends Error {
  record: PackInstallRecord;
  constructor(message: string, record: PackInstallRecord) {
    super(message);
    this.name = "PackInstallError";
    this.record = record;
  }
}

/** The basename of a pack-relative path. */
function base(relPath: string): string {
  return relPath.split("/").pop() ?? relPath;
}

/** A pack is third-party data, so every relative path is re-checked here even
 *  though the engine already classified it. The engine and the client can be
 *  different versions — a browser talking to an older engine is a supported
 *  configuration — and this one is cheap. */
function isSafeRelPath(relPath: string): boolean {
  if (!relPath || relPath.startsWith("/") || relPath.includes("\\")) return false;
  const parts = relPath.split("/");
  if (parts.some((p) => p === "" || p === "." || p === "..")) return false;
  return true;
}

/** What a pack install would do, without doing any of it. */
export function planPackInstall(
  pack: BackportPack,
  target: InstalledTitle,
  packRoot: string,
  existing: {
    fakelib: string[];
    sceModule: string[];
    gamePrx: string[];
    sceSys: string[];
    eboot: boolean;
  },
): PackInstallPlan {
  if (!pack.isPack) {
    throw new Error("This folder has no fakelib/, so it is not a backport pack");
  }
  if (!target.source) {
    throw new Error(`${target.titleId} has no source folder on the console`);
  }
  const root = packRoot.replace(/\/+$/, "");
  const dest = target.source.replace(/\/+$/, "");
  const stashDir = `${PACK_STASH_ROOT}/${target.titleId}`;

  const copies: PackCopy[] = [];
  const stashed: { live: string; stash: string }[] = [];

  const add = (file: PackFile, consolePath: string, alreadyThere: boolean, stashName: string) => {
    if (!isSafeRelPath(file.relPath)) {
      throw new Error(`Unsafe path in pack: ${file.relPath}`);
    }
    copies.push({ from: `${root}/${file.relPath}`, to: consolePath, size: file.size });
    if (alreadyThere) stashed.push({ live: consolePath, stash: `${stashDir}/${stashName}` });
  };

  for (const lib of pack.libraries) {
    const name = base(lib.relPath);
    add(lib, `${dest}/fakelib/${name}`, existing.fakelib.includes(name), `fakelib__${name}`);
  }
  for (const mod of pack.sceModules) {
    const name = base(mod.relPath);
    add(mod, `${dest}/sce_module/${name}`, existing.sceModule.includes(name), `sce_module__${name}`);
  }
  // Found by inspecting a second real pack: the first one shipped none of
  // these, and a classifier built from that single sample dropped all 26 of
  // the next pack's engine plugins on the floor.
  for (const plugin of pack.gamePrx) {
    const name = base(plugin.relPath);
    add(plugin, `${dest}/prx/${name}`, existing.gamePrx.includes(name), `prx__${name}`);
  }
  for (const f of pack.sceSys) {
    const name = base(f.relPath);
    add(f, `${dest}/sce_sys/about/${name}`, existing.sceSys.includes(name), `sce_sys_about__${name}`);
  }
  // The eboot goes LAST for the same reason the library flow patches the SDK
  // last: if anything before it fails, the title still has its original eboot
  // and still runs exactly as it did. Swapping the eboot first and then
  // failing leaves a game that launches into a crash.
  if (pack.eboot) {
    add(pack.eboot, `${dest}/eboot.bin`, existing.eboot, "eboot.bin");
  }

  return {
    target,
    packRoot: root,
    copies,
    stashed,
    stashDir,
    titleMismatch: !!pack.titleIdHint && pack.titleIdHint !== target.titleId,
    totalBytes: copies.reduce((sum, c) => sum + c.size, 0),
  };
}

export async function applyPackInstall(
  plan: PackInstallPlan,
  transport: PackTransport,
): Promise<PackInstallRecord> {
  const record: PackInstallRecord = {
    titleId: plan.target.titleId,
    targetSource: plan.target.source,
    packRoot: plan.packRoot,
    copiedPaths: [],
    stashed: [],
    stashDir: plan.stashDir,
    complete: false,
  };
  try {
    if (plan.stashed.length > 0) await transport.mkdirConsole(plan.stashDir);
    // Back everything up BEFORE writing anything, so a failure midway through
    // the copies still has every original available to undo from.
    for (const item of plan.stashed) {
      await transport.copyConsole(item.live, item.stash);
      record.stashed.push(item);
    }
    for (const dir of ["fakelib", "sce_module", "prx", "sce_sys", "sce_sys/about"]) {
      await transport.mkdirConsole(`${plan.target.source}/${dir}`);
    }
    for (const copy of plan.copies) {
      await transport.uploadHost(copy.from, copy.to);
      record.copiedPaths.push(copy.to);
    }
  } catch (error) {
    throw new PackInstallError(
      error instanceof Error ? error.message : String(error),
      record,
    );
  }
  record.complete = true;
  return record;
}

export async function undoPackInstall(
  record: PackInstallRecord,
  transport: PackTransport,
): Promise<void> {
  const restored = new Set(record.stashed.map((s) => s.live));
  // Anything we added that was not displacing an original is pure addition —
  // remove it. Reverse order so the eboot goes first, mirroring install.
  for (const path of [...record.copiedPaths].reverse()) {
    if (!restored.has(path)) await transport.remove(path);
  }
  // Put the originals back over the top of what we wrote.
  for (const item of record.stashed) {
    await transport.remove(item.live);
    await transport.copyConsole(item.stash, item.live);
  }
  // Non-semantic: once the originals are back, a leftover backup folder is
  // recoverable and far safer than reporting a failed Undo.
  try {
    await transport.remove(record.stashDir);
  } catch {
    /* keep the backup rather than fail the undo */
  }
}

File: engine/crates/ps5upload-core/src/backport_pack.rs
//! Recognising a community "backport pack" on disk.
//!
//! A pack is the format backports are actually distributed in, and it is much
//! more than a library set: alongside `fakelib/` it ships an `eboot.bin` that
//! has ALREADY been patched to the backport SDK pair, replacement
//! `sce_module/` modules, and the game's own engine plugins.
//!
//! Classification is on the RELATIVE PATH, never the basename. The corpus
//! importer used to decide "is this a library?" from the extension alone, and
//! every pack carries `.prx` files outside `fakelib/` — `sce_module/libc.prx`,
//! a `prx/` folder of engine plugins, `sce_sys/about/right.sprx`. Pointing that
//! importer at a pack folder silently produced a set mixing all of them
//! together, which is a combination no game has ever run.
//!
//! Packs vary in what they carry: some ship only `fakelib/`, `eboot.bin` and
//! `sce_module/`, others add dozens of engine plugins under `prx/`. Every
//! installable role below has to be recognised or those files are silently
//! left behind.

use std::path::{Path, PathBuf};

use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};

use crate::fakelibs::is_library_name;

/// What one file inside a pack is for.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PackRole {
    /// `fakelib/<name>.sprx` — the reusable library set, and the only part
    /// that belongs in the content-addressed corpus.
    Library,
    /// A pre-patched `eboot.bin` at the pack root.
    Eboot,
    /// `sce_module/<name>.prx` — replacement system modules installed beside
    /// the eboot. Emphatically NOT libraries.
    SceModule,
    /// `prx/<name>.prx` — the game's own engine plugins, re-signed by the pack
    /// author. Title-specific, so never corpus material, but they MUST be
    /// installed: a pack that ships them expects them to be there.
    GamePrx,
    /// `sce_sys/about/<name>` — shipped by every pack seen so far.
    SceSys,
    /// Everything else the pack happens to carry.
    Other,
}

impl PackRole {
    /// Does installing a pack write this file to the console?
    pub fn is_installable(self) -> bool {
        !matches!(self, PackRole::Other)
    }
}

/// Classify one pack-relative path.
///
/// Rejects anything that tries to escape the pack root: a pack is a folder the
/// user downloaded from a third party, so `..` and absolute paths are hostile
/// input, not edge cases.
pub fn classify(rel_path: &str) -> PackRole {
    let norm = rel_path.replace('\\', "/");
    if norm.starts_with('/') || norm.split('/').any(|c| c == ".." || c == ".") {
        return PackRole::Other;
    }
    let parts: Vec<&str> = norm.split('/').filter(|p| !p.is_empty()).collect();
    match parts.as_slice() {
        // These folders are flat: a nested path under one is not something any
        // pack ships, and treating it as installable would let a crafted
        // archive place a name we never vetted.
        ["fakelib", name] if is_library_name(name) => PackRole::Library,
        ["eboot.bin"] => PackRole::Eboot,
        ["sce_module", name] if is_library_name(name) => PackRole::SceModule,
        ["prx", name] if is_library_name(name) => PackRole::GamePrx,
        ["sce_sys", "about", name] if !name.starts_with('.') => PackRole::SceSys,
        _ => PackRole::Other,
    }
}

/// A title id mentioned in a pack's folder name.
///
/// The packs seen in the wild carry no manifest — the folder name is the only
/// statement of which game they are for (e.g.
/// `[SITE]-FW 4xx PPSA19534 (v01.000.016) backport files`). That makes this a
/// HINT and never an authority: the caller must confirm against the title the
/// user actually picked before overwriting a 256 MB eboot.
pub fn title_id_hint(folder_name: &str) -> Option<String> {
    let bytes = folder_name.as_bytes();
    for start in 0..bytes.len() {
        let rest = &folder_name[start..];
        if rest.len() < 9 {
            break;
        }
        let (head, tail) = rest.split_at(4);
        if !(head.eq_ignore_ascii_case("PPSA") || head.eq_ignore_ascii_case("CUSA")) {
            continue;
        }
        let digits: String = tail.chars().take(5).collect();
        if digits.len() == 5 && digits.chars().all(|c| c.is_ascii_digit()) {
            return Some(format!("{}{}", head.to_ascii_uppercase(), digits));
        }
    }
    None
}

/// One file found inside a pack.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PackFile {
    /// Path relative to the pack root, with `/` separators.
    pub rel_path: String,
    pub size: u64,
}

/// What a folder on disk turned out to contain.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct PackContents {
    pub libraries: Vec<PackFile>,
    pub eboot: Option<PackFile>,
    pub sce_modules: Vec<PackFile>,
    pub game_prx: Vec<PackFile>,
    pub sce_sys: Vec<PackFile>,
    pub other: Vec<PackFile>,
    /// Title id read out of the folder name, when it states one.
    pub title_id_hint: Option<String>,
}

impl PackContents {
    /// A pack with no `fakelib/` is not a pack. Saying so beats importing
    /// nothing and reporting success, which is how a user ends up believing
    /// the corpus grew when it did not.
    pub fn is_pack(&self) -> bool {
        !self.libraries.is_empty()
    }

    /// Everything an install would write, in the order it would write it.
    pub fn installable(&self) -> Vec<&PackFile> {
        let mut out: Vec<&PackFile> = Vec::new();
        out.extend(self.libraries.iter());
        out.extend(self.sce_modules.iter());
        out.extend(self.game_prx.iter());
        out.extend(self.sce_sys.iter());
        // The eboot goes last — see `planPackInstall` on the client for why.
        out.extend(self.eboot.iter());
        out
    }

    pub fn total_bytes(&self) -> u64 {
        self.installable().iter().map(|f| f.size).sum()
    }
}

/// Walk `dir` and sort what is in it into pack roles.
///
/// Descends at most two levels: every role lives at the root, one directory
/// down, or (for `sce_sys/about/`) two. An unbounded walk over a folder the
/// user pointed at is a way to spend minutes stat-ing a games library by
/// mistake.
pub fn inspect(dir: &Path) -> Result<PackContents> {
    if !dir.is_dir() {
        return Err(anyhow!("{} is not a folder", dir.display()));
    }
    let mut out = PackContents {
        title_id_hint: dir
            .file_name()
            .and_then(|n| n.to_str())
            .and_then(title_id_hint),
        ..Default::default()
    };

    let mut stack: Vec<(PathBuf, String, usize)> = vec![(dir.to_path_buf(), String::new(), 0)];
    while let Some((path, prefix, depth)) = stack.pop() {
        let entries =
            std::fs::read_dir(&path).map_err(|e| anyhow!("reading {}: {e}", path.display()))?;
        for entry in entries.flatten() {
            let Some(name) = entry.file_name().to_str().map(str::to_owned) else {
                continue;
            };
            // AppleDouble sidecars and .DS_Store ride along with anything
            // copied off a Mac and are never part of the pack.
            if name.starts_with('.') {
                continue;
            }
            let rel = if prefix.is_empty() {
                name.clone()
            } else {
                format!("{prefix}/{name}")
            };
            let Ok(meta) = entry.metadata() else { continue };
            if meta.is_dir() {
                // Two levels: `sce_sys/about/` is the deepest role there is.
                if depth < 2 {
                    stack.push((entry.path(), rel, depth + 1));
                }
                continue;
            }
            let file = PackFile {
                rel_path: rel.clone(),
                size: meta.len(),
            };
            match classify(&rel) {
                PackRole::Library => out.libraries.push(file),
                PackRole::Eboot => out.eboot = Some(file),
                PackRole::SceModule => out.sce_modules.push(file),
                PackRole::GamePrx => out.game_prx.push(file),
                PackRole::SceSys => out.sce_sys.push(file),
                PackRole::Other => out.other.push(file),
            }
        }
    }
    for v in [
        &mut out.libraries,
        &mut out.sce_modules,
        &mut out.game_prx,
        &mut out.sce_sys,
        &mut out.other,
    ] {
        v.sort_by(|a, b| a.rel_path.cmp(&b.rel_path));
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_fakelib_contents_are_libraries() {
        // Every one of these ends in .sprx/.prx, so an extension-only check
        // would sweep them all into the library set.
        assert_eq!(classify("fakelib/libSceAgc.sprx"), PackRole::Library);
        assert_eq!(classify("sce_module/libc.prx"), PackRole::SceModule);
        assert_eq!(
            classify("Engine.Render.Core2.PlatformPs5.retail.prx"),
            PackRole::Other
        );
        assert_eq!(
            classify("SP/Engine.Render.Core2.PlatformPs5.retail.prx"),
            PackRole::Other
        );
    }

    #[test]
    fn game_plugins_under_prx_are_installed_not_dropped() {
        // Some packs ship dozens of these; they must be installed, not
        // dropped as unrecognised.
        assert_eq!(classify("prx/akdelay.prx"), PackRole::GamePrx);
        assert_eq!(classify("prx/masteringsuite.prx"), PackRole::GamePrx);
        assert!(PackRole::GamePrx.is_installable());
        // Still not corpus material — title-specific, not a shared library.
        assert_ne!(classify("prx/akdelay.prx"), PackRole::Library);
    }

    #[test]
    fn sce_sys_about_is_reached_two_levels_down() {
        assert_eq!(classify("sce_sys/about/right.sprx"), PackRole::SceSys);
        assert!(PackRole::SceSys.is_installable());
    }

    #[test]
    fn the_root_eboot_is_recognised() {
        assert_eq!(classify("eboot.bin"), PackRole::Eboot);
        // Only at the root: a nested one belongs to something else.
        assert_eq!(classify("SP/eboot.bin"), PackRole::Other);
    }

    #[test]
    fn windows_separators_classify_the_same() {
        assert_eq!(classify("fakelib\\libSceAgc.sprx"), PackRole::Library);
        assert_eq!(classify("prx\\akdelay.prx"), PackRole::GamePrx);
    }

    #[test]
    fn traversal_and_absolute_paths_are_never_installable() {
        // A pack is third-party data; these must not reach the console.
        assert_eq!(classify("../fakelib/libSceAgc.sprx"), PackRole::Other);
        assert_eq!(classify("fakelib/../../etc/passwd"), PackRole::Other);
        assert_eq!(classify("/fakelib/libSceAgc.sprx"), PackRole::Other);
        assert_eq!(classify("fakelib/sub/libSceAgc.sprx"), PackRole::Other);
        assert_eq!(classify("prx/../../etc/passwd"), PackRole::Other);
        assert!(!PackRole::Other.is_installable());
    }

    #[test]
    fn appledouble_sidecars_are_not_libraries() {
        assert_eq!(classify("fakelib/._libSceAgc.sprx"), PackRole::Other);
        assert_eq!(classify("sce_sys/about/._right.sprx"), PackRole::Other);
    }

    #[test]
    fn the_title_id_comes_out_of_the_folder_name() {
        assert_eq!(
            title_id_hint("[DLPSGAME.COM]-FW 4xx PPSA19534 (v01.000.016) backport files")
                .as_deref(),
            Some("PPSA19534")
        );
        // The second pack uses underscores rather than spaces.
        assert_eq!(
            title_id_hint("[DLPSGAME.COM]-FW_4xx_PPSA29343_Beast_of_Reincarnation_01_000_000")
                .as_deref(),
            Some("PPSA29343")
        );
        assert_eq!(
            title_id_hint("CUSA07842 backport").as_deref(),
            Some("CUSA07842")
        );
        assert_eq!(title_id_hint("no title here"), None);
        // Too few digits is not a title id.
        assert_eq!(title_id_hint("PPSA123"), None);
    }

    #[test]
    fn a_folder_without_fakelib_is_not_a_pack() {
        let contents = PackContents {
            sce_modules: vec![PackFile {
                rel_path: "sce_module/libc.prx".into(),
                size: 10,
            }],
            ..Default::default()
        };
        assert!(!contents.is_pack());
    }

    #[test]
    fn inspect_sorts_a_real_pack_layout() {
        // The richer pack shape: every role present at once.
        let dir = std::env::temp_dir().join(format!("packtest-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        for sub in ["fakelib", "sce_module", "prx", "sce_sys/about"] {
            std::fs::create_dir_all(dir.join(sub)).unwrap();
        }
        std::fs::write(dir.join("fakelib/libSceAgc.sprx"), b"aaa").unwrap();
        std::fs::write(dir.join("fakelib/libkernel.sprx"), b"bb").unwrap();
        std::fs::write(dir.join("sce_module/libc.prx"), b"cccc").unwrap();
        std::fs::write(dir.join("prx/akdelay.prx"), b"dd").unwrap();
        std::fs::write(dir.join("prx/akgain.prx"), b"e").unwrap();
        std::fs::write(dir.join("sce_sys/about/right.sprx"), b"ff").unwrap();
        std::fs::write(dir.join("eboot.bin"), b"ggggg").unwrap();
        std::fs::write(dir.join("Engine.Render.retail.prx"), b"h").unwrap();

        let got = inspect(&dir).unwrap();
        assert!(got.is_pack());
        assert_eq!(got.libraries.len(), 2);
        assert_eq!(got.sce_modules.len(), 1);
        assert_eq!(got.game_prx.len(), 2, "prx/ plugins must be picked up");
        assert_eq!(got.sce_sys.len(), 1, "sce_sys/about is two levels down");
        assert_eq!(got.eboot.as_ref().unwrap().size, 5);
        // The loose engine .prx at the root is NOT installable.
        assert_eq!(got.other.len(), 1);
        // 3+2 libs, 4 module, 2+1 prx, 2 sce_sys, 5 eboot = 19; the root .prx
        // is excluded because nothing installs it.
        assert_eq!(got.total_bytes(), 19);
        // Eboot last, so a failure never leaves a swapped eboot behind.
        assert_eq!(got.installable().last().unwrap().rel_path, "eboot.bin");
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn inspect_rejects_a_file_or_missing_folder() {
        assert!(inspect(Path::new("/definitely/not/here")).is_err());
    }
}
